### ISSUE TITLE
Performance under load with custom workflows

### ISSUE THREAD

**Issue by LachlannRattray** [umbraco/Umbraco.Forms.Issues#97]
Hello
 
We have been load testing our Umbraco environment using MS load tests in visual studio. It is a simple solution which is comprised of a few Umbraco forms.
The Umbraco form we are testing has a custom work-flow attached, which connects to an API and updates the underlying form record using the response of the API call.
 
The form is quite simple, containing roughly 28 different fields.
 
Our app is hosted on an Azure p1v2 web app, and the database has 50 DTUs (S2)

We are using Umbraco 7.12.3 & Umbraco Forms 7.0.3

We have found that when load testing users hitting the site over a period of 5 minutes that the performance is quite slow. 

For example 
 
**25 users**
The application slows down significantly with a load of 25 users hammering the application at the same time, but the application remains available and there are no failures reported. Average POST time is 7.62 seconds and the average GET time is 1.08 seconds. Average CPU utilization on the app service plan increased steadily, spiking just once under 90%. Database DTU (Data Throughput Unit) utilization reached 100%, indicating heavy load on the database

**100 users**
Under constant load of 100 users the application performs very poorly. Here, average POST time is 28.9 seconds, and average GET time is 11.9 seconds. Average CPU utilization on the app service plan reached just under 100%, with some relief in between peaks. DTU consumption spiked approaching 100% on a few occasions with a short drop in between

This shows that the infrastructure is not coping well with the volume of requests.
In addition there were 50 failed requests / tests. These were due to the server being overwhelmed with requests. These appear to be timeouts where the POST request timed out.
 
We would not expect the infrastructure to have difficulty coping with this number of requests. When we look at application insights we see there is a lot of WAIT time as tasks get stuck waiting for resources, and also there is a considerable spike in the CPU thread count.

In looking at Umbraco forms it was found that
•	Not saving the data to the database greatly improved performance
•	Not executing the custom workflow increased performance on the POST request. Infact, it was found that the workflow task was poor even it had essentially an empty body, i.e not doing anything.

Further investigation using JetBrains dottrace appears to show that the GetForm method calls GetForms which executes a clone of all the forms before using SingleOrDefault to pick the correct form

```
public Form GetForm(Guid id)
    {
      if (Configuration.GetSetting("DisableFormCaching") == true.ToString())
        return this.GetFromFile(id);
      Form form = this.GetAllForms().SingleOrDefault<Form>((Func<Form, bool>) (x => x.Id == id));
      if (form != null)
        return form;
      throw new NullReferenceException(string.Format("Unable to get Form with ID: '{0}' off disk or from the cache. Check the file exists on disk", (object) id));
    }

public IEnumerable<Form> GetAllForms()
    {
      return ApplicationContext.Current.ApplicationCache.RuntimeCache.GetCacheItem<IEnumerable<Form>>("Forms.FormStorage.All", new Func<IEnumerable<Form>>(((BaseFileStorage<Form>) this).GetAll), new TimeSpan?(TimeSpan.FromMinutes(10.0)), false, CacheItemPriority.Normal, (CacheItemRemovedCallback) null, (string[]) null).Clone<IEnumerable<Form>>();
    }
```

If I am reading this correctly then this clone will be performed non-lazily so all forms are copied prior to being filtered, before the SingleOrDefault method gets invoked.

```
public static T Clone<T>(this T obj)
    {
      IFormatter formatter = (IFormatter) new BinaryFormatter();
      formatter.SurrogateSelector = (ISurrogateSelector) new SurrogateSelector();
      formatter.SurrogateSelector.ChainSelector((ISurrogateSelector) new NonSerialiazableTypeSurrogateSelector());
      MemoryStream memoryStream = new MemoryStream();
      formatter.Serialize((Stream) memoryStream, (object) obj);
      memoryStream.Position = 0L;
      return (T) formatter.Deserialize((Stream) memoryStream);
    }
```

In summary it seems like there are some performance issues with Umbraco forms in both writing to the DB and also in using custom work-flows. Perhaps optimizing the process above will help but also looking at how and when the system locks threads might be an idea?

Hope this is all clear, FYI I am posting this on behalf of a colleague who has been working on this issue so I may get him to jump in on any subsequent discussion.

Cheers

**Comment by bergmania** [comment]
Hi @LachlannRattray.

Thanks for reporting. The finding about cloning is fixed in the next release.

The other suggested fixes are not that simple for fix, so they should be created in separate issues, if this is still a problem after testing with this fix.

**Comment by LachlannRattray** [comment]
Thanks very much @bergmania this makes sense, do you know the timeline for the release of the next version of forms?

**Comment by bergmania** [comment]
@LachlannRattray Hopefully next tuesday

**Comment by NurhakKaya** [comment]
> Thanks very much @bergmania this makes sense, do you know the timeline for the release of the next version of forms?

Hello @LachlannRattray ; I wonder if you had a chance to check this performance fix, please?

According to [Umbraco 7.1.0(26 March 2019) release details,](https://our.umbraco.com/packages/developer-tools/umbraco-forms/) performance optimisations had been done for this issue, but I have just done an Umbraco Forms upgrade to 7.4.1 (Umbraco version 7.15.4 assembly: 1.0.7381.11453) and I see that performance is not good(page load times are more than 100 seconds or more on each Form page).

I have got 10 pages, on some pages there are some custom workflows, similar to your form, and on the fe website, I only render the form(apart from header and footer bits).

**Comment by LachlannRattray** [comment]
Hey @NurhakKaya  
I moved away from the project we had which showed these performance issues however I do not believe that it solved the issues we could see. From our further analysis there were many reasons that a very large form was performing badly. It might be worth looking at the version of forms compatible with v8 as I think there may be more perf improvements in that however I have not tried this.

**Comment by NurhakKaya** [comment]
Hello;

Just to update you that I can now confirm that this is not an Umbraco Forms bug and this was something to do with my project's custom membership provider and my wrong web.config setting for the UmbracoMembershipProvider. I have added a detailed answer for the problem and fix [on this Umbraco forum page,](https://our.umbraco.com/forum/umbraco-pro/contour/71915-forms-taking-a-long-time-to-load#comment-324174) hope it helps.

Thanks,
